Potential Control Flow Integrity Violation in wasm_runtime_malloc_internal

## Describe the Bug

The `wasm_runtime_malloc_internal` function may be vulnerable to control flow integrity violations, potentially due to the misuse or mismanagement of the `malloc_func` function pointer.

`iwasm --version`: iwasm 1.2.3

`uname -a`: Linux node2.khan22-169985.randtest.emulab.net 5.15.0-56-generic #62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

## Steps to reproduce:
1. Compile the code with CFISAN enabled:
```
    -DCMAKE_C_COMPILER_RANLIB=/usr/bin/ranlib \
    -DCMAKE_CXX_COMPILER_RANLIB=/usr/bin/ranlib \
    -DCMAKE_C_COMPILER_AR=/usr/bin/ar \
    -DCMAKE_CXX_COMPILER_AR=/usr/bin/ar \
    -DCMAKE_C_COMPILER_LAUNCHER=/usr/local/bin/afl-clang-fast \
    -DCMAKE_C_COMPILER_AR=/usr/bin/ar  \
    -DCMAKE_CXX_COMPILER_LAUNCHER=/usr/local/bin/afl-clang-fast++ \
    -DCMAKE_C_FLAGS="-fsanitize=cfisan-g" \
    -DCMAKE_CXX_FLAGS="-fsanitize=cfisan -g"
```
2. Save the `illegal.wat` (Remove .txt extension)
3. Install and build WABT toolkit (`wat2wasm` from WABT can be used to convert `illegal.wat` into `illegal.wasm` using: `wat2wasm illegal.wat -o illegal.wasm`. I have also added `illegal.wasm` that obviates the need to use `wat2wasm`)
4. Run `iwasm illegal.wasm`
5. Observe the `Illegal instruction`.

## Files that cause the issue:
[illegal.wat.txt](https://github.com/bytecodealliance/wasm-micro-runtime/files/12811314/illegal.wat.txt)

[illegal.wasm.txt](https://github.com/bytecodealliance/wasm-micro-runtime/files/12811315/illegal.wasm.txt)


## Observed Behavior:
Potential control flow integrity violation if `malloc_func` is not properly managed or if it's tampered with.
```
iwasm illegal.wasm 
Illegal instruction
```

## Expected behavior:
The `malloc_func` should always point to a valid memory allocation function and should not be overwritten or tampered with by external or unexpected means.

## GDB Backtrace

```
Reading symbols from ../targets/wasm-micro-runtime/product-mini/platforms/linux/build/iwasm...
(gdb) break main
Breakpoint 1 at 0x9ae0: file /root/wasmoi/targets/wasm-micro-runtime/product-mini/platforms/linux/../posix/main.c, line 540.
(gdb) run illegal.wasm 
Starting program: /root/wasmoi/targets/wasm-micro-runtime/product-mini/platforms/linux/build/iwasm illegal.wasm
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, main (argc=2, argv=0x7fffffffe418) at /root/wasmoi/targets/wasm-micro-runtime/product-mini/platforms/linux/../posix/main.c:540
540	{
(gdb) c
Continuing.

Program received signal SIGILL, Illegal instruction.
0x0000555555567cb9 in wasm_runtime_malloc_internal (size=32) at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_memory.c:177
177	        return malloc_func(size);
(gdb) bt
#0  0x0000555555567cb9 in wasm_runtime_malloc_internal (size=32) at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_memory.c:177
#1  wasm_runtime_malloc (size=32) at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_memory.c:253
#2  0x000055555555e3dd in register_natives (n_native_symbols=52, call_conv_raw=false, module_name=<optimized out>, native_symbols=<optimized out>)
    at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_native.c:337
#3  wasm_native_register_natives (n_native_symbols=52, module_name=<optimized out>, native_symbols=<optimized out>)
    at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_native.c:377
#4  wasm_native_init () at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_native.c:573
#5  wasm_runtime_env_init () at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_runtime_common.c:432
#6  wasm_runtime_full_init (init_args=<optimized out>) at /root/wasmoi/targets/wasm-micro-runtime/core/iwasm/common/wasm_runtime_common.c:672
#7  main (argc=<optimized out>, argv=<optimized out>) at /root/wasmoi/targets/wasm-micro-runtime/product-mini/platforms/linux/../posix/main.c:871
```

### The code whose execution firstly found the issue (`cat illegal.wat`):

```
(module (memory $mem 1) (global $lift_1 (mut i32) (i32.const 346))
   (table 0 funcref)
   (elem (i32.const 0))
   (func
    $main
    (export "_main")
    (result i32)
    (local $i32_storage i32)
    (local $i64_storage i64)
    (local $f32_storage f32)
    (local $f64_storage f64)
    i32.const
    468
    i32.load16_s
    offset=13
    align=1
    global.get
    $lift_1
    i32.and
    global.set
    $lift_1
    global.get
    $lift_1
    i32.popcnt))
```

## Additional information

A combination of AFL++ and Wasmlike, an Xsmith-based random program generator produced the snippet of code that found the issue. [Xsmith Project](https://www.flux.utah.edu/project/xsmith)

@khagankhan Could you upload the whole command to build iwasm? I used the below command:
```bash
AFL_USE_ASAN=1 CC=/usr/bin/afl-clang-fast CXX=/usr/bin/afl-clang-fast++ cmake .. \
-DCMAKE_C_COMPILER_RANLIB=/usr/bin/ranlib \
-DCMAKE_CXX_COMPILER_RANLIB=/usr/bin/ranlib \
-DCMAKE_C_COMPILER_AR=/usr/bin/ar \
-DCMAKE_CXX_COMPILER_AR=/usr/bin/ar \
-DCMAKE_C_COMPILER_LAUNCHER=/usr/bin/afl-clang-fast \
-DCMAKE_CXX_COMPILER_LAUNCHER=/usr/bin/afl-clang-fast++ \
-DCMAKE_C_FLAGS="-fsanitize=cfisan-g" \
-DCMAKE_CXX_FLAGS="-fsanitize=cfisan-g"
```
But got error:
```bash
  CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
  The C compiler
    "/usr/bin/afl-clang-fast"
  is not able to compile a simple test program.
  ....
  clang: error: unsupported argument 'cfisan-g' to option 'fsanitize='
```

--- a/product-mini/platforms/posix/main.c
+++ b/product-mini/platforms/posix/main.c
@@ -891,6 +891,10 @@ main(int argc, char *argv[])
     init_args.mem_alloc_option.pool.heap_size = sizeof(global_heap_buf);
 #else
     init_args.mem_alloc_type = Alloc_With_Allocator;
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    /* Set user data for the allocator is needed */
+    /* init_args.mem_alloc_option.allocator.user_data = user_data; */
+#endif
     init_args.mem_alloc_option.allocator.malloc_func = malloc_func;
     init_args.mem_alloc_option.allocator.realloc_func = realloc_func;
     init_args.mem_alloc_option.allocator.free_func = free_func;

--- a/product-mini/platforms/posix/main.c
+++ b/product-mini/platforms/posix/main.c
@@ -464,7 +464,37 @@ moudle_destroyer(uint8 *buffer, uint32 size)
 
 #if WASM_ENABLE_GLOBAL_HEAP_POOL != 0
 static char global_heap_buf[WASM_GLOBAL_HEAP_SIZE] = { 0 };
+#else
+static void *
+malloc_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
 #endif
+    unsigned int size)
+{
+    return malloc(size);
+}
+
+static void *
+realloc_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
+#endif
+    void *ptr, unsigned int size)
+{
+    return realloc(ptr, size);
+}
+
+static void
+free_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
+#endif
+    void *ptr)
+{
+    free(ptr);
+}
+#endif /* end of WASM_ENABLE_GLOBAL_HEAP_POOL */
 
 #if WASM_ENABLE_STATIC_PGO != 0
 static void
@@ -861,9 +891,9 @@ main(int argc, char *argv[])
     init_args.mem_alloc_option.pool.heap_size = sizeof(global_heap_buf);
 #else
     init_args.mem_alloc_type = Alloc_With_Allocator;
-    init_args.mem_alloc_option.allocator.malloc_func = malloc;
-    init_args.mem_alloc_option.allocator.realloc_func = realloc;
-    init_args.mem_alloc_option.allocator.free_func = free;
+    init_args.mem_alloc_option.allocator.malloc_func = malloc_func;
+    init_args.mem_alloc_option.allocator.realloc_func = realloc_func;
+    init_args.mem_alloc_option.allocator.free_func = free_func;
 #endif
 
 #if WASM_ENABLE_FAST_JIT != 0

--- a/product-mini/platforms/windows/main.c
+++ b/product-mini/platforms/windows/main.c
@@ -191,7 +191,37 @@ validate_env_str(char *env)
 
 #if WASM_ENABLE_GLOBAL_HEAP_POOL != 0
 static char global_heap_buf[WASM_GLOBAL_HEAP_SIZE] = { 0 };
+#else
+static void *
+malloc_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
 #endif
+    unsigned int size)
+{
+    return malloc(size);
+}
+
+static void *
+realloc_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
+#endif
+    void *ptr, unsigned int size)
+{
+    return realloc(ptr, size);
+}
+
+static void
+free_func(
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    void *user_data,
+#endif
+    void *ptr)
+{
+    free(ptr);
+}
+#endif /* end of WASM_ENABLE_GLOBAL_HEAP_POOL */
 
 #if WASM_ENABLE_MULTI_MODULE != 0
 static char *
@@ -450,9 +480,9 @@ main(int argc, char *argv[])
     init_args.mem_alloc_option.pool.heap_size = sizeof(global_heap_buf);
 #else
     init_args.mem_alloc_type = Alloc_With_Allocator;
-    init_args.mem_alloc_option.allocator.malloc_func = malloc;
-    init_args.mem_alloc_option.allocator.realloc_func = realloc;
-    init_args.mem_alloc_option.allocator.free_func = free;
+    init_args.mem_alloc_option.allocator.malloc_func = malloc_func;
+    init_args.mem_alloc_option.allocator.realloc_func = realloc_func;
+    init_args.mem_alloc_option.allocator.free_func = free_func;
 #endif
 
 #if WASM_ENABLE_JIT != 0

--- a/product-mini/platforms/windows/main.c
+++ b/product-mini/platforms/windows/main.c
@@ -480,6 +480,10 @@ main(int argc, char *argv[])
     init_args.mem_alloc_option.pool.heap_size = sizeof(global_heap_buf);
 #else
     init_args.mem_alloc_type = Alloc_With_Allocator;
+#if WASM_MEM_ALLOC_WITH_USER_DATA != 0
+    /* Set user data for the allocator is needed */
+    /* init_args.mem_alloc_option.allocator.user_data = user_data; */
+#endif
     init_args.mem_alloc_option.allocator.malloc_func = malloc_func;
     init_args.mem_alloc_option.allocator.realloc_func = realloc_func;
     init_args.mem_alloc_option.allocator.free_func = free_func;

